File: VMTranslator/cpp/VMTranslator.cpp
/*
* Program: The Hack Translator (vm code to assembly code)
* Author : Xi (Ronnie) Chen
* Date   : 06/04/2022
* Usage  : prompt> VMTranslator ProgName.vm
*/

#include <iostream>
#include <string>
#include <filesystem>

#include <boost/program_options.hpp>

#include "Parser.hpp"
#include "CodeWriter.hpp"

namespace fs = std::filesystem;
namespace po = boost::program_options;

// ============================================================ //

int main(int argc, char** argv)
{
	// -------------------------------------------------------- //
	// argparse: get the name of the input source file
	// -------------------------------------------------------- //

	po::options_description desc{ "Hack VM Translator" };

	desc.add_options()
		("help", "produce help message")
		("file", po::value<std::string>(), "hack vm code file")
	;

	po::positional_options_description p;
	p.add("file", -1);
	
	po::variables_map vm;        
	po::store(po::command_line_parser(argc, argv).options(desc).positional(p).run(), vm);
	po::notify(vm);

	if (vm.count("help")) {
		std::cout << desc << std::endl;
	    return 0;
	}

	if (vm.count("file") == 0) {
		std::cout << "Please provide a hack vm code file!" << std::endl;
		std::cout << "Usage: VMTranslator.exe Prog.vm" << std::endl;
		return 1;
	}

	fs::path vm_file_path{ vm["file"].as<std::string>() };
	fs::path asm_file_path{ vm_file_path };
	asm_file_path.replace_extension("asm");

	// -------------------------------------------------------- //
	// vm file processing
	// -------------------------------------------------------- //

	try {
		// init
		Parser parser{ vm_file_path };
		CodeWriter writer{ asm_file_path };

		// parsing
		std::cout << "Parsing [" << vm_file_path.string() << "]... ";
		while (parser.has_more_lines()) {
			parser.advance();
			if (!parser.is_curr_cmd_valid()) {
				continue;
			}
			auto cmd_t = parser.command_type();
			if (cmd_t == CmdType::C_ARITHMETIC) {
				writer.write_arithmetic(parser.arg1());
			}
			else if (cmd_t == CmdType::C_PUSH || cmd_t == CmdType::C_POP) {
				writer.write_push_pop(cmd_t, parser.arg1(), parser.arg2());
			}
			else {
				// TODO
			}
		}
		std::cout << "done..." << std::endl;

		std::cout << "File written to: " << asm_file_path.string() << std::endl;
	}
	catch (const std::exception& e) {
		std::cout << e.what() << std::endl;
		return 1;
	}

	// -------------------------------------------------------- //

	return 0;
}

// ============================================================ //

File: VMTranslator/VMTranslator.py
# %% Import Libs

import argparse

from pathlib import Path

from Parser import Parser
from MyTypes import CmdType
from CodeWriter import CodeWriter

# %% vm file processing

def process_vm_file(vm_file_path: Path, writer: CodeWriter):

    print(f"Parsing [{vm_file_path}]... ", end="")

    parser = Parser(vm_file_path)
    writer.set_file_name(vm_file_path.name)

    while parser.has_more_lines():
        parser.advance()
        if not parser.is_curr_cmd_valid():
            continue
        cmd_t = parser.command_type()
        if cmd_t == CmdType.C_ARITHMETIC:
            writer.write_arithmetic(parser.arg1())
        elif (cmd_t == CmdType.C_PUSH) or (cmd_t == CmdType.C_POP):
            writer.write_push_pop(cmd_t, parser.arg1(), parser.arg2())
        elif cmd_t == CmdType.C_LABEL:
            writer.write_label(parser.arg1())
        elif cmd_t == CmdType.C_GOTO:
            writer.write_goto(parser.arg1())
        elif cmd_t == CmdType.C_IF:
            writer.write_if(parser.arg1())
        elif cmd_t == CmdType.C_FUNCTION:
            writer.write_function(parser.arg1(), parser.arg2())
        elif cmd_t == CmdType.C_RETURN:
            writer.write_return()
        elif cmd_t == CmdType.C_CALL:
            writer.write_call(parser.arg1(), parser.arg2())
        else:
            raise Exception(f"Unrecognized command type: [{cmd_t.name}]")

    parser.close()

    print("done...")

# %% main

def _main():

    # argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("file")
    options = parser.parse_args()

    # path processing
    input_path = Path(options.file)
    if input_path.is_dir():
        vm_files = input_path.glob("*.vm")
        asm_file_path = (input_path/input_path.name).with_suffix(".asm")
    else:
        vm_files = [input_path]
        asm_file_path = input_path.with_suffix(".asm")

    writer = CodeWriter(asm_file_path)

    for vm_file in vm_files:
        process_vm_file(vm_file, writer)

    writer.close()

    print(f"File written to: {asm_file_path}")

# %% run main

if __name__ == "__main__":
    _main()
